store oauth tokens per client app as sorted set, new endpoint to get …

dencoded opened the pull request:
changes for https://github.com/TykTechnologies/tyk-analytics/issues/533

- store in Redis list of tokens per oauth-client with using sorted set
- use token expire date as a score so we can request range of tokens not expired as for now
- remove range with already expired tokens to free up space for other keys
- having token list stored as a sorted set we can easily add pagination if required (just moving range for pages)
- also, added new endpoint `GET /oauth/clients/{apiID}/{keyName}/tokens` to get list of not expired tokens per API and oauth-client

reply format example:
```
[
    {
        "code": "5a7d110be6355b0c071cc339327563cb45174ae387f52f87a80d2496",
        "expires": 1518158407
    },
    {
        "code": "5a7d110be6355b0c071cc33988884222b0cf436eba7979c6c51d6dbd",
        "expires": 1518158594
    },
    {
        "code": "5a7d110be6355b0c071cc33990bac8b5261041c5a7d585bff291fec4",
        "expires": 1518158638
    },
    {
        "code": "5a7d110be6355b0c071cc339a66afe75521f49388065a106ef45af54",
        "expires": 1518159792
    }
]
```

buger commented:
Nice!  

Lets do token expiration time configurable via tyk.conf, if it is 0, do not remove. Does it make sense?

No comments on code itself, but we need tests for this functionality.

dencoded commented:
@buger yep, we need to test it carefully. also, I am still not sure about `RPCStorageHandler` - there are method stubs there at the moment. Seems like these kind of API calls can be sent only via dashboard API to management gateway or directly to management gateway (not to slave), so I guess we are good.

According to token expiration - looks like we already have it configurable via config and there is no way to get 0 there, correct me if I am missing something.

Initially it gets set here https://github.com/TykTechnologies/tyk/blob/master/vendor/github.com/lonelycode/osin/authorize.go#L213 which I believe have initial hardcoded value `3600`


Then it gets overwritten with value (if set) from `config.Global.OauthTokenExpire` here https://github.com/TykTechnologies/tyk/pull/1465/files#diff-2264b1a6175f98974544452eae683529R619

I didn't touch this peace btw.

buger commented:
Sorry for misunderstanding, In this case, I meant time after "expired" tokens gets removed.

dencoded commented:
@buger PTAL - I've added:
- new config key `oauth_token_expired_retain_period ` to specify retain period for expired tokens stored in redis sorted sets
- test for new endpoint `GET /oauth/clients/{apiID}/{keyName}/tokens`
- added changes to make api ID optional in `POST /oauth/clients/create`  - these changes make endpoint more generic, if we provide API ID it works the same it was before, if we don't provide API ID it uses policy's access rights and enumerates APIs from there setting the same newly created oauth-client

I've tested it with case when we have requested oauth-client for some policy ID which contains several APIs - after request got approved we can get tokens for any API from that policy using the same oauth client credentials

buger commented on oauth_manager_test.go:
```diff
@@ -70,10 +78,10 @@ const oauthDefinition = `{
 	// Ensure all the correct ahndlers are in place
 	loadAPIEndpoints(muxer)
-	manager := addOAuthHandlers(spec, muxer)
+	spec.OAuthManager = addOAuthHandlers(spec, muxer)
```
👍

dencoded replied:
ok, I will work on this

buger commented:
@dencoded pls also fix conflicts

dencoded commented:
@buger I've submitted changes in tests. please note a couple of changes in instrumentation:
- `(s *tykTestServer) Do(tc test.TestCase) (*http.Response, error)` is not following redirects anymore (useful to tests 3XX replies)
- `(s *tykTestServer) Run(t *testing.T, testCases ...test.TestCase) (*http.Response, error)` - uses copy of response for asserts and returns unclosed original response to be read in test logic if needed

buger commented:
Looks good! 

Added RPC/Hybrid support notice into phase 2 https://github.com/TykTechnologies/product/issues/19
